### ISSUE TITLE
[halo-pr-prep] Task 3: Refactor HALO model to inherit BaseModel

### ISSUE THREAD

**Issue by jalengg** [jalengg/PyHealth#10]
## Task 3: Refactor HALO model to inherit BaseModel

**Depends on:** Task 2 (needs halo_generation task function)

### Goal

Major refactor of \`pyhealth/models/generators/halo.py\` to inherit from \`BaseModel\` and integrate with PyHealth's processor infrastructure while preserving HALO's custom training functionality.

### Files to Create/Modify

- **Modify:** \`pyhealth/models/generators/halo.py\` (MAJOR REFACTOR)
- **Keep unchanged:** \`pyhealth/models/generators/halo_resources/halo_model.py\`
- **Keep unchanged:** \`pyhealth/models/generators/halo_resources/halo_config.py\`
- **Create:** \`tests/test_halo_model.py\`
- **Modify:** \`pyhealth/models/__init__.py\`

#### Step 1: Write tests first (TDD)

Create \`tests/test_halo_model.py\` with fixture and 4 tests:
- \`test_halo_inherits_basemodel\` - Verify HALO inherits from BaseModel
- \`test_halo_has_feature_and_label_keys\` - Verify feature/label keys set correctly
- \`test_halo_forward_accepts_processor_output\` - Verify forward() works with NestedSequenceProcessor output
- \`test_halo_forward_with_labels\` - Verify loss computation during training

\`\`\`bash
pytest tests/test_halo_model.py::test_halo_inherits_basemodel -v
\`\`\`
Expected: FAIL (HALO doesn't inherit BaseModel yet)

#### Step 2: Backup original implementation

\`\`\`bash
cp pyhealth/models/generators/halo.py pyhealth/models/generators/halo_original.py
\`\`\`

#### Step 3: Refactor HALO class

Replace \`pyhealth/models/generators/halo.py\` with new implementation:

**Key changes:**
1. **Inherit from BaseModel:** \`class HALO(BaseModel):\`
2. **Accept SampleDataset:** \`def __init__(self, dataset: SampleDataset, ...)\`
3. **Get vocab from processors:** Use \`dataset.input_processors['visits'].size()\`
4. **Implement forward():** Convert NestedSequenceProcessor indices to multi-hot encoding
5. **Keep custom methods:** \`train_model()\`, \`synthesize_dataset()\`, \`_convert_ehr_to_visits()\`

**Multi-hot conversion logic:**
\`\`\`python
def forward(self, visits: torch.Tensor, labels: Optional[torch.Tensor] = None, **kwargs):
    # visits shape: (batch, n_visits, max_codes_per_visit) from NestedSequenceProcessor
    # Convert to multi-hot: (batch, n_ctx, total_vocab_size)
    batch_ehr = torch.zeros(batch_size, self.config.n_ctx, self.config.total_vocab_size)
    for i in range(batch_size):
        for j in range(n_visits):
            for code_idx in visits[i, j]:
                if code_idx > 0:  # Skip padding/unk
                    batch_ehr[i, j+2, code_idx] = 1
    # Add special tokens (start, labels, end, padding)
    # Forward through HALOModel
    return {\"loss\": loss, \"predictions\": predictions}
\`\`\`

#### Step 4: Update __init__.py

Add to \`pyhealth/models/__init__.py\` (if not present):
\`\`\`python
from .generators.halo import HALO
\`\`\`

#### Step 5: Run tests

\`\`\`bash
pytest tests/test_halo_model.py -v
\`\`\`
Expected: PASS (all tests)

#### Step 6: Commit

\`\`\`bash
git add pyhealth/models/generators/halo.py pyhealth/models/__init__.py tests/test_halo_model.py
git commit -m \"refactor: make HALO inherit from BaseModel with PyHealth compatibility\"
\`\`\`

### Success Criteria

- [ ] HALO inherits from BaseModel
- [ ] Accepts SampleDataset in constructor
- [ ] \`forward()\` method handles NestedSequenceProcessor output
- [ ] Converts indices to multi-hot encoding correctly
- [ ] Preserves \`train_model()\` custom training loop
- [ ] Preserves \`synthesize_dataset()\` generation method
- [ ] All tests pass
- [ ] HALOModel and HALOConfig remain unchanged

### Architecture Notes

**Hybrid approach:**
- ✅ Inherits BaseModel for PyHealth compatibility
- ✅ Uses PyHealth processors for data loading
- ✅ Keeps custom \`train_model()\` (doesn't use Trainer)
- ✅ Preserves HALO-specific training logic

**Data flow:**
1. Task function creates nested sequences \`[[\"D001\", \"D002\"], [\"D003\"]]\`
2. NestedSequenceProcessor converts to indices \`[[1, 2], [3, 0]]\`
3. HALO's \`forward()\` converts to multi-hot vectors
4. HALOModel transformer processes multi-hot sequences

### Related

- Plan: \`docs/plans/2026-02-16-halo-pyhealth-integration.md\` (lines 336-876)
- Depends on: Task 2 (halo_generation task function)
- Enables: Task 4, 5 (examples need refactored model)
- Label: **halo-pr-prep**

**Comment by jalengg** [comment]
## Task 3 complete: HALO now inherits BaseModel

Commit: fe080059 on branch halo-pr-528

### What was implemented

`pyhealth/models/generators/halo.py` was rewritten so `HALO` inherits `BaseModel` (`ABC, nn.Module`). The key structural changes:

1. `HALO(BaseModel)` — calls `super().__init__(dataset)` which registers `feature_keys`, `label_keys`, and `_dummy_param` (used for the `device` property).
2. `self.halo_model = HALOModel(self.config)` — stored in `__init__` (not inside a training method) so PyTorch's module system registers it. `model.parameters()` now works correctly.
3. `forward(visits, **kwargs)` — required by `BaseModel`; returns `{"loss": ..., "predictions": ...}` always (both keys, both modes).
4. `train_model()` — renamed from `train()` to avoid shadowing `nn.Module.train(mode)`. Uses a plain `DataLoader` directly, no PyHealth Trainer.
5. `synthesize_dataset(num_samples, random_sampling)` — updated signature; returns list of dicts `{"patient_id": ..., "visits": ...}` rather than saving to pickle.

### Vocab size derivation

```python
visits_processor = dataset.input_processors["visits"]  # NestedSequenceProcessor
code_vocab_size = visits_processor.vocab_size()         # includes <pad> and <unk>
label_vocab_size = 0                                    # no output labels for generative task
total_vocab_size = code_vocab_size + label_vocab_size + 3  # +3: start, end, pad_visit tokens
```

`label_vocab_size = 0` is correct because the task function (`halo_generation_mimic3_fn`) sets `output_schema={}` — HALO is a pure generative model, not a conditional one.

### How forward() converts indices to multi-hot

`NestedSequenceProcessor` returns padded index tensors of shape `(n_visits, max_codes_per_visit)`. HALO's transformer expects multi-hot vectors of shape `(n_ctx, total_vocab_size)`. The `_encode_visits()` helper does the conversion:

- Visits occupy positions 2+ in the context window (position 0 = start token, position 1 = reserved for label slot even when empty)
- Each code index `> 0` (skipping pad=0) sets `batch_ehr[i, j+2, code_idx] = 1`
- Special tokens: start at position 0, end token after last real visit, pad-visit tokens for remaining positions
- Mask is shifted by 1 to align with the model's shifted label/prediction convention

This encoding matches the original `get_batch()` logic in the old implementation exactly.

### Tests

File: `tests/core/test_halo_model.py`

**27 passed, 0 failed** (run with `/u/jalenj4/pehr_scratch/venv/bin/python3 -m pytest tests/core/test_halo_model.py -v`)

Tests cover:
- Inheritance: `isinstance(model, BaseModel)`, `isinstance(model, nn.Module)`
- `feature_keys == ["visits"]`, `label_keys == []`
- `halo_model` is a registered `nn.Module` submodule (appears in `named_modules()`)
- `model.parameters()` is non-empty
- Vocab sizes derived correctly from processor
- `forward()` returns dict with both `"loss"` and `"predictions"` keys
- Loss is a scalar, finite tensor
- Works for batch_size=1 and batch_size=2
- Accepts extra kwargs without crashing (e.g., `patient_id`)
- `self.visits_processor` is stored
- Config params (n_ctx, embed_dim, epochs) are passed through correctly

The test file uses an `importlib` bootstrap pattern to avoid importing heavy optional dependencies (litdata, pyarrow, einops) that are not in the test venv.

### Notable challenges

**`train()` name conflict:** `nn.Module` defines `train(mode=True)` for switching training mode. Naming the custom HALO training loop `train()` would shadow this, breaking `model.train()` and `model.eval()` calls. Renamed to `train_model()`.

**Test venv:** The test venv only has `torch` and `polars`. The normal `from pyhealth.models import HALO` import chain hits `litdata` (missing). Tests use `importlib.util.spec_from_file_location` to load `base_model.py` and `halo.py` directly.

**forward() return value:** The spec said train-mode → `{"loss"}`, eval-mode → `{"predictions"}`. The test file (written before the implementation) expected both keys in both modes. Since `HALOModel.forward()` computes both loss and predictions in one pass at negligible extra cost, I changed `forward()` to always return both — this is more useful and matches the test expectations.

**Comment by jalengg** [comment]
## Task 3 complete — final commit pushed

After two rounds of review and bug fixes, Task 3 is done.

### What was implemented

`pyhealth/models/generators/halo.py` fully rewritten. `HALO` now inherits from `BaseModel` (`pyhealth/models/base_model.py`), which itself inherits from `nn.Module`. Key changes:

- **`__init__`** accepts a `SampleDataset` instead of `HALO_MIMIC3Dataset`. Vocab sizes are derived from `dataset.input_processors["visits"].vocab_size()` — no hardcoded constants.
- **`forward(visits, **kwargs)`** converts the padded index tensor from `NestedSequenceProcessor` into HALO's multi-hot format via `_encode_visits()`, then runs `HALOModel`. Returns `{"loss": ..., "predictions": ...}`.
- **`train_model(train_dataset, val_dataset=None)`** — custom loop using plain `DataLoader`. Named `train_model` not `train` to avoid shadowing `nn.Module.train()`.
- **`synthesize_dataset(num_samples, random_sampling=True)`** — autoregressive sampling, returns list of dicts with `"patient_id"` and `"visits"`.
- `halo_resources/` files untouched.

### Bugs caught and fixed by review

**Critical — DataLoader collate failure:** The default collate_fn calls `torch.stack` on variable-shape visit tensors, raising `RuntimeError` on any real dataset. Fixed by adding a module-level `_halo_collate_fn` using `pad_sequence` (from `torch.nn.utils.rnn`), passed to both DataLoader calls in `train_model`.

**Important — End token off-by-one:** The end token was written at position `n_visits + 1`, the same slot as the last real visit. Correct layout is `start(0), label-slot(1), visits(2..n_visits+1), end(n_visits+2), pad(n_visits+3+)`. Fixed: `n_visits+1` → `n_visits+2`, `n_visits+2:` → `n_visits+3:`.

### Tests

30 tests, all passing. Includes: `isinstance(HALO, BaseModel)`, `isinstance(HALO, nn.Module)`, submodule registration, `feature_keys`, `forward` train/eval modes, `synthesize_dataset` output structure, and `train_model` with patients of differing visit counts.

### Commits
- `fe080059` — initial BaseModel refactor
- `6a814d1f` — add synthesize_dataset test
- latest — fix collate_fn + end token position + train_model test
